cclib/method/ddec.py

```diff
@@ -0,0 +1,220 @@
+            else:
+                self.chgdensity = electrondensity_spin(
+                    self.data,
+                    self.volume,
+                    [
+                        self.data.mocoeffs[0][: self.data.homos[0]],
+                        self.data.mocoeffs[1][: self.data.homos[1]],
+                    ],
+                )
+        # If charge densities are provided beforehand, log this information
+        # `Volume` object does not contain (nor rely on) information about the constituent atoms.
+        else:
+            self.logger.info("Using charge densities from the provided Volume object.")
+            self.chgdensity = self.volume
+            
+        # STEP 1
+        # Carry out step 1 of DDEC6 algorithm [Determining ion charge value]
+        # Refer to equations 49-57 in doi: 10.1039/c6ra04656h
+        self.calculate_refcharges()
+
+
+    def calculate_refcharges(self):
+        # Generator object to iterate over the grid
+        xshape, yshape, zshape = self.chgdensity.data.shape
+        atoms = len(self.data.atomnos)
+        indices = ((i, x, y, z) for i in range(atoms) for x in range(xshape) for y in range(yshape) for z in range(zshape))
+        
+        stockholder_w = numpy.zeros((atoms, xshape, yshape, zshape))
+        localized_w = numpy.zeros((atoms, xshape, yshape, zshape))
+        
+        for atomi, xindex, yindex, zindex in indices:
+            # Distance of the grid from atom grid
+            dist_r = self._cartesian_dist(self.data.atomcoords[-1][atomi], self.chgdensity.coordinates([xindex, yindex, zindex]))
+            closest_r_index = numpy.abs(self.radial_grid_r[atomi] - dist_r).argmin()
+            
+            stockholder_w[atomi][xindex][yindex][zindex] = self.proatom_density[atomi][closest_r_index]
+        
+        localized_w = numpy.power(stockholder_w, 4)
+        
+        stockholder_bigW = numpy.sum(stockholder_w, axis = 0)
+        localized_bigW = numpy.sum(localized_w, axis = 0)
+        
+        self.logger.info("Creating first reference charges.")
+        self.refcharges = numpy.zeros((atoms))
+        self._localizedcharges = numpy.zeros((atoms))
+        self._stockholdercharges = numpy.zeros((atoms))
+        
+        for atomi in range(atoms):
+            chargedensity = copy.deepcopy(self.chgdensity)
+            localized_weight = localized_w[atomi]/localized_bigW
+            chargedensity.data *= localized_weight
+            self._localizedcharges[atomi] = self.data.atomnos[atomi] - chargedensity.integrate()
+            
+            chargedensity = copy.deepcopy(self.chgdensity)
+            stockholder_weight = stockholder_w[atomi]/stockholder_bigW
+            chargedensity.data *= stockholder_weight
+            self._stockholdercharges[atomi] = self.data.atomnos[atomi] - chargedensity.integrate()
+            
+            self.refcharges[atomi] = self._stockholdercharges[atomi] / 3.0 + 2.0 * self._localizedcharges[atomi] / 3.0
```

Comment:
can you add some parentheses to group this a bit.

Reply:
I've added some parentheses and a comment here. Thanks!